File: leetcode/32_Longest_Valid_Parentheses.py
import unittest

class Solution:
    def longestValidParentheses(self, s: str) -> int:
        n = len(s)
        stack, valids = [], [False] * n
        for i in range(n):
            if s[i] == '(':
                stack.append(i)
            elif len(stack) > 0:
                valids[stack.pop()] = True
                valids[i] = True
        cnt, max_cnt = 0, 0
        for valid in valids:
            if valid:
                cnt += 1
                max_cnt = cnt if cnt > max_cnt else max_cnt
            else:
                cnt = 0
        return max_cnt

class TestLongestValidParentheses(unittest.TestCase):
    def test_1(self):
        self.assertEqual(Solution().longestValidParentheses("(()"), 2)    

    def test_2(self):
        self.assertEqual(Solution().longestValidParentheses(")()())"), 4)    

    def test_3(self):
        self.assertEqual(Solution().longestValidParentheses(""), 0)    

    def test_4(self):
        self.assertEqual(Solution().longestValidParentheses("()(()"), 2)    

File: leetcode/32_Longest_Valid_Parentheses.rs
struct Solution {}

impl Solution {
    pub fn longest_valid_parentheses(s: String) -> i32 {
        let s: Vec<char> = s.chars().collect();
        let n = s.len();
        let mut stack: Vec<usize> = vec![];
        let mut valids = vec![false; n];
        for i in 0..n {
            if s[i] == '(' {
                stack.push(i);
            } else if stack.len() > 0 {
                valids[stack.pop().unwrap()] = true;
                valids[i] = true;
            } 
        }
        let (mut cnt, mut max_cnt) = (0, 0);
        for valid in valids{
            if valid {
                cnt += 1;
                if cnt > max_cnt { max_cnt = cnt; }
            } else {
                cnt = 0;
            }
        }
        max_cnt
    }
}

#[cfg(test)]
mod longest_valid_parentheses_test {
    use super::*;

    #[test]
    fn test_1() {
        assert_eq!(Solution::longest_valid_parentheses("(()".to_string()), 2);
    }

    #[test]
    fn test_2() {
        assert_eq!(Solution::longest_valid_parentheses(")()())".to_string()), 4);
    }

    #[test]
    fn test_3() {
        assert_eq!(Solution::longest_valid_parentheses("".to_string()), 0);
    }
}
